File: C0012S/BOJ1932.java
/*
1932. Silver 1 - 정수 삼각형

    시간 제한	    메모리 제한        제출        정답	      맞힌 사람	    정답 비율
    2 초	    128 MB           82612	    47657     35958	         59.297%


    문제
                    7
                  3   8
                8   1   0
              2   7   4   4
            4   5   2   6   5

        위 그림은 크기가 5인 정수 삼각형의 한 모습이다.
        맨 위층 7부터 시작해서 아래에 있는 수 중 하나를 선택하여 아래층으로 내려올 때, 이제까지 선택된 수의 합이 최대가 되는 경로를 구하는 프로그램을 작성하라. 아래층에 있는 수는 현재 층에서 선택된 수의 대각선 왼쪽 또는 대각선 오른쪽에 있는 것 중에서만 선택할 수 있다.
        삼각형의 크기는 1 이상 500 이하이다. 삼각형을 이루고 있는 각 수는 모두 정수이며, 범위는 0 이상 9999 이하이다.


    입력
        첫째 줄에 삼각형의 크기 n(1 ≤ n ≤ 500)이 주어지고, 둘째 줄부터 n+1번째 줄까지 정수 삼각형이 주어진다.


    출력
        첫째 줄에 합이 최대가 되는 경로에 있는 수의 합을 출력한다.


    예제 입력 1
        5
        7
        3 8
        8 1 0
        2 7 4 4
        4 5 2 6 5
    예제 출력 1
        30


    알고리즘 분류
        다이나믹 프로그래밍
*/


// 메모리 : 27760KB
// 시간 : 316ms
// 코드 길이 : 1686B
// 정답

package C0012S;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ1932 {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer token;
        int n = Integer.parseInt(bf.readLine()); // 삼각형의 크기 n (1 ≤ n ≤ 500)

        int triangle[][] = new int[n][n]; // 크기가 n인 정수 삼각형의 수들을 저장한 배열
        int dp[][] = new int[n][n]; // dp[x][y] : (x, y) 위치까지 올 수 있는 경로들의 수의 합 중 최댓값
        for (int r = 0; r < n; r++) {
            token = new StringTokenizer(bf.readLine());

            for (int c = 0; c <= r; c++) {
                triangle[r][c] = Integer.parseInt(token.nextToken());

                if (r == 0) { // 초깃값(dp[0][0])에 저장될 값은 삼각형의 첫 번째 수
                    dp[r][c] = triangle[r][c];
                }
                else {
                    if (c == 0) { // 해당 행의 첫 번째 원소일 경우
                        dp[r][c] = dp[r - 1][c] + triangle[r][c];
                    }
                    else {
                        dp[r][c] = Math.max(dp[r - 1][c - 1], dp[r - 1][c]) + triangle[r][c];
                    }
                }
            }
        }

        int maxSum = 0; // 합이 최대가 되는 경로에 있는 수의 합
        for (int m = 0; m < n; m++) { // 2 차원 배열 dp의 마지막 행에서 최댓값 찾기
            maxSum = Math.max(maxSum, dp[n - 1][m]);
        }

        System.out.println(maxSum);
    }
}

File: gyeoul/BOJ1932.kt
/*
1932번: 정수 삼각형
입력받은 삼각형을 다음 형태로 저장한다
0 0
0 7 0
0 3 8 0
0 8 1 0 0
0 2 7 4 4 0
0 4 5 2 6 5 0
0 0 0 0 0 0 0 0
각 줄의 첫번째와 마지막 숫자를 제외한 나머지 숫자의
도달할 수 있는 경로는 [i-1][j-1], [i-1][j]
즉 자신의 바로 위의 숫자와 그 왼쪽 숫자만이 도달할 수 있는 유일한 경로이다
두 값중 큰 값을 선택하여 현재 위치의 값을 더한 뒤 맨 마지막 줄에서
같은 계산을 시도하여 그중 가장 큰 값이 이 삼각형에서 선택할 수 있는
가장 큰 합이다.
*/
import java.util.StringTokenizer

class BOJ1932 {
    fun main() = with(System.`in`.bufferedReader()) {
        val bw = System.out.bufferedWriter()
        var st = StringTokenizer(readLine())
        val n = st.nextToken().toInt()
        val arr = ArrayList<IntArray>()
        for (i in 0..n + 1) {
            val now = IntArray(i + 2)
            if (i in 1..n) {
                st = StringTokenizer(readLine())
                for (j in 1..i) {
                    now[j] = st.nextToken().toInt()
                }
            }
            arr.add(now)
        }
        for (i in 1..n + 1) {
            for (j in 1..arr[i].size - 2) {
                arr[i][j] += Math.max(arr[i - 1][j - 1], arr[i - 1][j])
            }
        }
        bw.write("${arr[n + 1].max()}")
        bw.flush()
    }
}

